File: built/Profiler.lua
 --[[/*--------------------------------------------------------------------
    Copyright (C) 2014 Johnny C. Lam.
    See the file LICENSE.txt for copying permission.
--------------------------------------------------------------------*/]]--
local _addonName, _addon = ...
define(_addonName, _addon, "Profiler", {"Ovale", "AceConfig-3.0", "AceConfigDialog-3.0", 'Localization', "LibTextDump-1.0", './Options', 'stub'}, function (__exports, Ovale_1, AceConfig_3_0_1, AceConfigDialog_3_0_1, Localization_1, LibTextDump_1_0_1, Options_1, stub_1)
    local OVALE = "Ovale";
    local OvaleProfiler = (function () 
        local OvaleProfiler = {}
        OvaleProfiler.constructor = function (this)
            this.timestamp = stub_1.debugprofilestop();
            this.stack = {};
            this.stackSize = 0;
            this.timeSpent = {};
            this.timesInvoked = {};
            this.addons = {};
            this.profilingOutput = nil;
            this.options = {
                name= OVALE .. " " .. Localization_1.L["Profiling"],
                type= "group",
                args= {
                    profiling= {
                        name= Localization_1.L["Profiling"],
                        type= "group",
                        args= {
                            modules= {
                                name= Localization_1.L["Modules"],
                                type= "group",
                                inline= true,
                                order= 10,
                                args= {},
                                get= function (info)
                                    local name = info[stub_1.len(info) + 1];
                                    local value = Ovale_1.ovale.db.global.profiler[name + 1];
                                    return (value ~= nil);
                                end,
                                set= function (info, value)
                                    value = value or nil;
                                    local name = info[stub_1.len(info) + 1];
                                    Ovale_1.ovale.db.global.profiler[name + 1] = value;
                                    if (value) then 
                                        this.EnableProfiling(name);
                                    else 
                                        this.DisableProfiling(name);
                                    end
                                end,
                            },
                            reset= {
                                name= Localization_1.L["Reset"],
                                desc= Localization_1.L["Reset the profiling statistics."],
                                type= "execute",
                                order= 20,
                                func= function () this.ResetProfiling(); end,
                            },
                            show= {
                                name= Localization_1.L["Show"],
                                desc= Localization_1.L["Show the profiling statistics."],
                                type= "execute",
                                order= 30,
                                func= function ()
                                    this.profilingOutput.Clear();
                                    local s = this.GetProfilingInfo();
                                    if (s) then 
                                        this.profilingOutput.AddLine(s);
                                        this.profilingOutput.Display();
                                    end
                                end,
                            },
                        },
                    },
                },
            };
            this.array = {};
        end;
        OvaleProfiler.DoNothing = function (this)
        end;
        OvaleProfiler.StartProfiling = function (this, tag)
            local newTimestamp = stub_1.debugprofilestop();
             --[[// Attribute the time spent up to this call to the previous function.]]--
            if (this.stackSize > 0) then 
                local delta = newTimestamp - this.timestamp;
                local previous = this.stack[this.stackSize + 1];
                local timeSpent = this.timeSpent[previous + 1] or 0;
                timeSpent = timeSpent + delta;
                this.timeSpent[previous + 1] = timeSpent;
            end
             --[[// Add the current function to the call stack.]]--
            this.timestamp = newTimestamp;
            this.stackSize = this.stackSize + 1;
            this.stack[this.stackSize + 1] = tag;
            local timesInvoked = this.timesInvoked[tag + 1] or 0;
            timesInvoked = timesInvoked + 1;
            this.timesInvoked[tag + 1] = timesInvoked;
        end;
        OvaleProfiler.StopProfiling = function (this, tag)
            if (this.stackSize > 0) then 
                local currentTag = this.stack[this.stackSize + 1];
                if (currentTag == tag) then 
                    local newTimestamp = stub_1.debugprofilestop();
                    local delta = newTimestamp - this.timestamp;
                    local timeSpent = this.timeSpent[currentTag + 1] or 0;
                    timeSpent = timeSpent + delta;
                    this.timeSpent[currentTag + 1] = timeSpent;
                    this.timestamp = newTimestamp;
                    this.stackSize = this.stackSize - 1;
                end
            end
        end;
        OvaleProfiler.OnInitialize = function (this)
            local appName = this.GetName();
            AceConfig_3_0_1.default:RegisterOptionsTable(appName, this.options);
            AceConfigDialog_3_0_1.default:AddToBlizOptions(appName, Localization_1.L["Profiling"], OVALE);
        end;
        OvaleProfiler.OnEnable = function (this)
            if (not this.profilingOutput) then 
                this.profilingOutput = LibTextDump_1_0_1.default:New(OVALE .. " - " .. Localization_1.L["Profiling"], 750, 500);
            end
        end;
        OvaleProfiler.OnDisable = function (this)
            this.profilingOutput:Clear();
        end;
        OvaleProfiler.RegisterProfiling = function (this, addon, name)
            name = name or addon:GetName();
            this.addons[name + 1] = addon;
            this.options.args.profiling.args.modules.args[name] = {
                name= name,
                desc= stub_1.format(Localization_1.L["Enable profiling for the %s module."], name),
                type= "toggle",
            };
            this.DisableProfiling(name);
        end;
        OvaleProfiler.EnableProfiling = function (this, name)
            local addon = this.addons[name + 1];
            if (addon) then 
                addon.StartProfiling = this.StartProfiling;
                addon.StopProfiling = this.StopProfiling;
            end
        end;
        OvaleProfiler.DisableProfiling = function (this, name)
            local addon = this.addons[name + 1];
            if (addon) then 
                addon.StartProfiling = this.DoNothing;
                addon.StopProfiling = this.DoNothing;
            end
        end;
        OvaleProfiler.ResetProfiling = function (this)
            for local tag,_ in ipairs(this.timeSpent) do tag = tag - 1 
                do
                this.timeSpent[tag + 1] = nil;
                end
            end
            for local tag,_ in ipairs(this.timesInvoked) do tag = tag - 1 
                do
                this.timesInvoked[tag + 1] = nil;
                end
            end
        end;
        OvaleProfiler.GetProfilingInfo = function (this)
            if (stub_1.next(this.timeSpent)) then 
                 --[[// Calculate the width needed to print out the times invoked.]]--
                local width = 1;
                local tenPower = 10;
                for local key,_ in ipairs(this.timesInvoked) do key = key - 1 
                    do
                    local timesInvoked = this.timesInvoked[key + 1];
                    while (timesInvoked > tenPower)  do
                        do
                        width = width + 1;
                        tenPower = tenPower * 10;
                        end
                    end
                    end
                end
                stub_1.wipe(this.array);
                local formatString = stub_1.format("    %%08.3fms: %%0%dd (%%05f) x %%s", width);
                for local tag,_ in ipairs(this.timeSpent) do tag = tag - 1 
                    do
                    local timeSpent = this.timeSpent[tag + 1];
                    local timesInvoked = this.timesInvoked[tag + 1];
                    stub_1.tinsert(this.array, stub_1.format(formatString, timeSpent, timesInvoked, timeSpent / timesInvoked, tag));
                    end
                end
                if (stub_1.next(this.array)) then 
                    tsort(this.array);
                    local now = stub_1.GetTime();
                    stub_1.tinsert(this.array, 1, stub_1.format("Profiling statistics at %f:", now));
                    return stub_1.tconcat(this.array, "\n");
                end
            end
        end;
        OvaleProfiler.DebuggingInfo = function (this)
            Ovale_1.ovale:Print("Profiler stack size = %d", this.stackSize);
            local index = this.stackSize;
            while (index > 0 and this.stackSize - index < 10)  do
                do
                local tag = this.stack[index + 1];
                Ovale_1.ovale:Print("    [%d] %s", index, tag);
                index = index - 1;
                end
            end
        end;
        return OvaleProfiler;
    end)();
    __exports.ovaleProfiler = Ovale_1.ovale:NewModule("OvaleProfiler", __new(OvaleProfiler));
    local actions = {
        profiling= {
            name= Localization_1.L["Profiling"],
            type= "execute",
            func= function ()
                local appName = exports.ovaleProfiler:GetName();
                AceConfigDialog_3_0_1.default:SetDefaultSize(appName, 800, 550);
                AceConfigDialog_3_0_1.default:Open(appName);
            end,
        },
    };
     --[[// Insert actions into OvaleOptions.]]--
    for local k,_ in ipairs(actions) do k = k - 1 
        do
        Options_1.ovaleOptions.options.args.actions.args[k + 1] = actions[k + 1];
        end
    end
     --[[// Add a global data type for debug options.]]--
    Options_1.ovaleOptions.defaultDB.global = Options_1.ovaleOptions.defaultDB.global or {};
    Options_1.ovaleOptions.defaultDB.global.profiler = {};
    Options_1.ovaleOptions:RegisterOptions(OvaleProfiler);
end);

File: src/Profiler.ts
/*--------------------------------------------------------------------
    Copyright (C) 2014 Johnny C. Lam.
    See the file LICENSE.txt for copying permission.
--------------------------------------------------------------------*/

/*
	Every time a new function is entered and exited, debugprofilestop() is called and the time between
	the two timestamps is calculated and attributed to that function.
*/

import { ovale } from "Ovale";
import AceConfig from "AceConfig-3.0";
import AceConfigDialog from "AceConfigDialog-3.0"
import { L } from 'Localization';
import LibTextDump from "LibTextDump-1.0";
import { ovaleOptions } from './Options';
import { format, ipairs, next, tconcat, tinsert, sort, wipe, GetTime, len, debugprofilestop } from 'stub';
var OVALE = "Ovale";

class ProfiledModule {
	private DoNothing(tag: string) {}

	public StartProfiling: (tag:string) => void;
	public StopProfiling: (tag: string) => void;

	constructor(private profiler: OvaleProfiler) {
	}

	private DoStartProfiling(tag: string) {
		this.profiler.StartProfiling(tag);
	}

	private DoStopProfiling(tag: string){
		this.profiler.StopProfiling(tag);
	}

	public Enable(){
		this.StartProfiling = this.DoStartProfiling;
		this.StopProfiling = this.DoStopProfiling;
	}

	public Disable(){
		this.StartProfiling = this.DoNothing;
		this.StopProfiling = this.DoNothing;
	}
}

class OvaleProfiler implements ModuleEvents {
	addons: LuaDictionary<ProfiledModule> = {}

	profilingOutput:TextDump = null
	
	options = {
		name: OVALE + " " + L["Profiling"],
		type: "group",
		args: {
			profiling: {
				name: L["Profiling"],
				type: "group",
				args: {
					modules: {
						name: L["Modules"],
						type: "group",
						inline: true,
						order: 10,
						args: {},
						get: (info: LuaTable<string>) => {
							let name = info[len(info)]
							let value = ovale.db.global.profiler[name]
							return (value != null)
						},
						set: (info: LuaTable<string>, value: string) => {
							value = value || null;
							let name = info[len(info)]
							ovale.db.global.profiler[name] = value
							if (value) {
								this.EnableProfiling(name)
							} else {
								this.DisableProfiling(name)
							}
						},
					},
					reset: {
						name: L["Reset"],
						desc: L["Reset the profiling statistics."],
						type: "execute",
						order: 20,
						func: () => { this.ResetProfiling() },
					},
					show: {
						name: L["Show"],
						desc: L["Show the profiling statistics."],
						type: "execute",
						order: 30,
						func: () => {
							this.profilingOutput.Clear();
							let s = this.GetProfilingInfo();
							if (s) {
								this.profilingOutput.AddLine(s)
								this.profilingOutput.Display()
							}
						},
					},
				},
			},
		},
	}

	OnInitialize() {
		let appName = "OvaleProfiler";
		AceConfig.RegisterOptionsTable(appName, this.options)
		AceConfigDialog.AddToBlizOptions(appName, L["Profiling"], OVALE)
		
		var actions:any = {
			profiling: {
				name: L["Profiling"],
				type: "execute",
				func: function() {
					AceConfigDialog.SetDefaultSize(appName, 800, 550)
					AceConfigDialog.Open(appName)
				},
			},
		}
		// Insert actions into OvaleOptions.
		for (var k in actions) {
			ovaleOptions.options.args.actions.args[k] = actions[k];
		}
				
		// Add a global data type for debug options.
		ovaleOptions.defaultDB.global = ovaleOptions.defaultDB.global || {}
		ovaleOptions.defaultDB.global.profiler = {}
		ovaleOptions.RegisterOptions(this);
	}

	OnEnable() {
		if (! this.profilingOutput) {
			this.profilingOutput = LibTextDump.New(OVALE + " - " + L["Profiling"], 750, 500)
		}
	}

	OnDisable() {
		this.profilingOutput.Clear()
	}

	RegisterProfiling(name:string) {
		const addon = new ProfiledModule(this);
		this.addons[name] = addon;
		this.options.args.profiling.args.modules.args[name] = {
			name: name,
			desc: format(L["Enable profiling for the %s module."], name),
			type: "toggle",
		}
		this.DisableProfiling(name);
		return addon;
	}

	EnableProfiling(name:string) {
		let addon = this.addons[name]
		if (addon) {
			addon.Enable();
		}
	}

	DisableProfiling(name:string) {
		let addon = this.addons[name]
		if (addon) {
			addon.Disable();
		}
	}
	timestamp = debugprofilestop()
	stack:LuaTable<string> = {}
	stackSize = 0
	timeSpent:LuaDictionary<number> = {}
	timesInvoked:LuaDictionary<number> = {}
		
	StartProfiling(tag: string) {
		let newTimestamp = debugprofilestop()

		// Attribute the time spent up to this call to the previous function.
		if (this.stackSize > 0) {
			let delta = newTimestamp - this.timestamp
			let previous = this.stack[this.stackSize]
			let timeSpent = this.timeSpent[previous] || 0
			timeSpent = timeSpent + delta
			this.timeSpent[previous] = timeSpent
		}

		// Add the current function to the call stack.
		this.timestamp = newTimestamp
		this.stackSize = this.stackSize + 1
		this.stack[this.stackSize] = tag
		let timesInvoked = this.timesInvoked[tag] || 0
		timesInvoked = timesInvoked + 1
		this.timesInvoked[tag] = timesInvoked
	}

	StopProfiling(tag: string) {
		if (this.stackSize > 0) {
			let currentTag = this.stack[this.stackSize]
			if (currentTag == tag){
				let newTimestamp = debugprofilestop()
				let delta = newTimestamp - this.timestamp
				let timeSpent = this.timeSpent[currentTag] || 0
				timeSpent = timeSpent + delta
				this.timeSpent[currentTag] = timeSpent
				this.timestamp = newTimestamp
				this.stackSize = this.stackSize - 1
			}
		}
	}	


	array:LuaTable<string> = {}

	GetProfilingInfo() {
		if (next(this.timeSpent)) {
			// Calculate the width needed to print out the times invoked.
			let width = 1
			let tenPower = 10
			for (let key in this.timesInvoked) {
				let timesInvoked = this.timesInvoked[key];
				while (timesInvoked > tenPower) {
					width = width + 1
					tenPower = tenPower * 10
				}
			}
		
			wipe(this.array)
			let formatString = format("    %%08.3fms: %%0%dd (%%05f) x %%s", width)
			for (let tag in this.timeSpent) {
				let timeSpent = this.timeSpent[tag];
				let timesInvoked = this.timesInvoked[tag]
				tinsert(this.array, format(formatString, timeSpent, timesInvoked, timeSpent / timesInvoked, tag))
			}
			if (next(this.array)) {
				sort(this.array)
				let now = GetTime()
				tinsert(this.array, 1, format("Profiling statistics at %f:", now))
				return tconcat(this.array, "\n")
			}
		}
		return null;
	}
	

	DebuggingInfo() {
		ovale.Print("Profiler stack size = %d", this.stackSize)
		let index = this.stackSize
		while (index > 0 && this.stackSize - index < 10) {
			let tag = this.stack[index]
			ovale.Print("    [%d] %s", index, tag)
			index = index - 1
		}
	}
	
	ResetProfiling() {
		for (let tag in this.timeSpent) {
			this.timeSpent[tag] = null;
		}

		for (let tag in this.timesInvoked) {
			this.timesInvoked[tag] = null;
		}
	}
}

export const ovaleProfiler = ovale.NewModule("OvaleProfiler", new OvaleProfiler());


